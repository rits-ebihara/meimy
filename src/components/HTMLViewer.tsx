import React, { Component } from 'react';
import { NativeSyntheticEvent, WebView, WebViewMessageEventData } from 'react-native';


interface IProps {
    html: string;
}

interface IState {
    height: number;
}

export class HtmlViewer extends Component<IProps, IState> {
    public constructor(props: IProps) {
        super(props);
        this.state = {
            height: 200,
        };
    }
    public render() {
        return (
            <WebView
                source={{ html: this.createHtml(this.props.html) }}
                style={{ height: this.state.height }}
                onMessage={this.onMessage}
            />
        );
    }
    private onMessage = (event: NativeSyntheticEvent<WebViewMessageEventData>) => {
        let height = Number(event.nativeEvent.data);
        if (!Number.isNaN(height)) {
            // height = height / PixelRatio.get();
            this.setState({
                height,
            });
        }
    }
    private createHtml = (html: string) => {
        const completedHtml = `<!DOCTYPE html>
        <html lang="en">
        <head>
            <meta charset="UTF-8">
            <meta name="viewport" content="width=device-width,initial-scale=1">
            <style>
                html, body {
                    font-size: 14pt;
                }
                #container {
                    position: absolute;
                    top: 0;
                    left: 0;
                    right: 0;
                    margin: 0;
                    padding: 0;
                }
                p {
                    margin: 0;
                    padding: 0;
                }
            </style>
            <script>
                window.addEventListener("load", () => {
                    setTimeout(() => {
                        const container = document.getElementById('container');
                        window.ReactNativeWebView.postMessage(container.clientHeight);
                    }, 100);
                });
            </script>
        </head>
        <body>
            <div id="container">
                ${html}
            </div>
        </body>
        </html>`;
        return completedHtml;
    }
}

export default HtmlViewer
