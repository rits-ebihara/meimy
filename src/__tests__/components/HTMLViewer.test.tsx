import Enzyme, { shallow } from 'enzyme';
import EnzymeAdapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import React from 'react';

import HTMLViewer from '../../components/HTMLViewer';

Enzyme.configure({ adapter: new EnzymeAdapter() });

const html = `
    <p>ほげ</p>
`;

describe('render', () => {
    test('init', () => {
        const wrapper = shallow(<HTMLViewer html={html} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});

describe('event', () => {
    test('onMessage', () => {
        const wrapper = shallow<HTMLViewer>(<HTMLViewer html={html} />);
        const instance = wrapper.instance();
        instance['onMessage']({
            nativeEvent: {
                data: '300',
            },
        } as any);
        expect(wrapper.state().height).toEqual(300);
    });
    test('onMessage - not number', () => {
        const wrapper = shallow<HTMLViewer>(<HTMLViewer html={html} />);
        const instance = wrapper.instance();
        instance['onMessage']({
            nativeEvent: {
                data: 'abc',
            },
        } as any);
        expect(wrapper.state().height).toEqual(200);
    });

});
