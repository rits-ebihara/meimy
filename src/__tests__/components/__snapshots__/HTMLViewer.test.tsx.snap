// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render init 1`] = `
<WebView
  cacheEnabled={true}
  onMessage={[Function]}
  originWhitelist={
    Array [
      "http://*",
      "https://*",
    ]
  }
  source={
    Object {
      "html": "<!DOCTYPE html>
        <html lang=\\"en\\">
        <head>
            <meta charset=\\"UTF-8\\">
            <meta name=\\"viewport\\" content=\\"width=device-width,initial-scale=1\\">
            <style>
                html, body {
                    font-size: 14pt;
                }
                #container {
                    position: absolute;
                    top: 0;
                    left: 0;
                    right: 0;
                    margin: 0;
                    padding: 0;
                }
                p {
                    margin: 0;
                    padding: 0;
                }
            </style>
            <script>
                window.addEventListener(\\"load\\", () => {
                    setTimeout(() => {
                        const container = document.getElementById('container');
                        window.ReactNativeWebView.postMessage(container.clientHeight);
                    }, 100);
                });
            </script>
        </head>
        <body>
            <div id=\\"container\\">
                
    <p>ほげ</p>

            </div>
        </body>
        </html>",
    }
  }
  style={
    Object {
      "height": 200,
    }
  }
  useSharedProcessPool={true}
  useWebKit={true}
/>
`;
